### PR TITLE
Refactor to use Dask future APIs in frontend only

### PR DESCRIPTION
**Pull request by ayushbaid**
Summary of the  refactored design:
1. Each module will implement the `apply` function, which should compute the output products as well as the metrics. 
2. There will be no use of dask inside any module
3. The modules for which `apply` can be run in parallel (frontend) will be using Dask's futures. This logic will reside in `gtsfm_runner_base`

### CODE REVIEW

**Comment by ayushbaid** [comment]
Some timing data from CI

| Dataset | This branch (mins) | Master (mins)|
| -------- | ------------------------ | ---------------------- |
| door-12 SIFT front end | 24 |  23 |
| door-12 deep front end | 61 | 58 |
| notre dame-20 deep front end | 91 | 51 | 

This does not look promising. Maybe the CI is not a good place to get benchmarking data?

**Comment by ayushbaid** [comment]
Local runs indicate a similar performance degradation. So I guess the serialization and deserialization is more expensive than loading network weights from disk again. But to run GTSFM on a cluster, this might be a performance penalty we have to take. 

@johnwlambert @akshay-krishnan

**Comment by ayushbaid** [comment]
This PR is one way of resolving #110

**Comment by akshay-krishnan** [comment]
We could discuss this in more detail in a meeting. In the meantime, would it be worth benchmarking on the cluster? @stepanyanhayk could you help with that?

**Comment by stepanyanhayk** [comment]
> 

Of course! I will run this on the cluster and share the benchmarks.

**Comment by ayushbaid** [comment]
Thanks!

I expect performance to be even worse because of the network latency.

**Comment by akshay-krishnan** [comment]
@ayushbaid  I wish we could profile specific components. For example, how long does detector-descriptor take on a single image on average when using the delayed vs the futures API? that would give us some more insight into what exactly causes the latency. this is definitely nontrivial on distributed workers, but have you tried on a single machine?

**Comment by ayushbaid** [comment]
I benchmarked the detector in #608 and it showed promising results.

File: gtsfm/common/image.py [hunk 1]
```diff
@@ -34,6 +34,10 @@ def width(self) -> int:
         """
         return self.value_array.shape[1]
 
+    @property
+    def shape(self) -> Tuple[int, int]:
```
Comment (akshay-krishnan) [inline gtsfm/common/image.py:38]: I added a shape property already, please pull from master

**Pull request closed without merging**